### ISSUE TITLE
Cache methods should take Snowflakeish instead of Snowflake

### ISSUE THREAD

**Issue by Tmpod** [nekokatt/hikari#197]
### Summary

Cache methods, both for retrieval and updating (in mutable cache), should have their snowflake arguments typed as `Snowflakeish` instead of just `Snowflake`.

### Problem

The API is currently inconsistent, since, for example, the REST components  are typed with `Snowflakeish`, while the cache is typed with `Snowflake`. It's also more convenient to be able to use whatever we have at hand from these three quite interchangeable formats, `int`, `str` and `Snowflake`. Additionally, it would also address the fact that currently the cache methods can take ints all fine, despite not being typed as such.

### Ideal implementation

Of course the first step would be to change the methods' signatures on the base classes, and on the implementations. Then I suppose one would have to change the implementations to cast the given Snowflakeish to just one of the three types, possibly Snowflake being the best candidate, since there would be no need to change the cache's collections' typing and also because this type offers more utility methods which are, well, useful :P  One downside would be memory consumption, which would naturally be higher in Snowflake than in the builtins.

### Checklist

- [x] I have searched the issue tracker and have made sure it's not a duplicate.

**Comment by Tmpod** [comment]
One problem that Espy just thought about in a little chat on DCA was that allowing `Snowflakeish` would mean allowing `str`, meaning tons of `int` casts everywhere which would introduce a fair amount of overhead. This union type works great in REST because sooner or later the ID will have to be cast into a string in order to be sent in the JSON payload, but this doesn't quite fit this cache issue.   
Not sure what the best course of action would be.

**Comment by nekokatt** [comment]
the eventual plan is to have model methods/properties to access this info in a more controlled way.

While I appreciate that this is not ideal for now, it should probably be best to manually cast to Snowflake before touching the cache, if you need MyPy compat. I will discuss with @FasterSpeeding the implications of changing the cache interface to allow int as well. I assume there isn't much that needs changing other than the typehints, since Snowflake inherits from int anyway, so the `__eq__` and friends, and `__hash__` overloads are interoperable.

I _do_ eventually have plans to deprecate the use of str in Snowflakeish, since unless you are talking directly to the Discord API on a lower level than that which we provide, then you shouldn't need to use str anyway. I am not aware of any part of hikari that requires this anymore. str is a relic from when the core REST client provided almost zero transformation to/from the API JSON response/request shape. I will look to make an issue for this.

As for Unique, which is the fourth shape of Snowflakeish, we would definitely need casts to support this I think, unless we ensured all Unique models interoped 100% with ints for their eq and hash overloads, which I am not overly keen on. Since the cache interface technically is for consumption by the rest of hikari rather than by users (just havent had time to get this implemented just yet), there is a general assumption that, for example, if you already have a User object, you dont need the cache to look up the same object you already have.

----

# TLDR

1. We should look into making the cache consume ints rather than snowflakes. This'll remain snowflake compatible but fit your use case a little more nicely.
2. We need to look into removing str from Snowflakeish, as it will reduce the faffing about and simplify internals. **This should be separate to this issue.** and **should be done ASAP while we still have freedom to do this in the alpha stage.** Issue for this is `<<insert ticket number here when anyone is at a PC to type this up>>`.

Of course, if you fancy a challenge; once @FasterSpeeding has responded to confirm my plan has no caveats, anyone can pick this up if you get to it before one of us.

**Comment by nekokatt** [comment]
@Tmpod and @FasterSpeeding, looks like this is no where near as simple as I would have hoped. Since the cache is mapping snowflakes onto models using the Snowflake type, it means we either have to implement a bunch of casting, find a way to trick mypy into thinking int and Snowflake are interchangable, remove the snowflake type entirely across the API, or abandon this idea completely. Simply changing the annotations led to around 145 new MyPy errors that were mostly not trivial to fix without larger scale changes.

I have removed the `str` type support though. That was fine to remove.

Thoughts?

**Comment by Tmpod** [comment]
Casting doesn't sound too good; tricking mypy could work but I'm not entirely sure what you intend to do, I'm not familiar with the tool; removing the snowflake type doesn't seem like an option tbh, it's useful to have that object, and since it's slotted and adds no fields, just functions it ends up having pretty much the same memory fingerprint as ints; abandoning the idea should be a last resort thing tbh.

What would be the impacts of having cache do lots of casting?   
What about tricking mypy? Do you have any idea in mind, and what would be the implications of fooling it just for this cache issue?

**Comment by nekokatt** [comment]
> What would be the impacts of having cache do lots of casting?

Each cast is a function call. If you have 2000 members in a guild, each with 5 roles, you are doing 10,000 casts for a single guild just on roles alone. Extrapolate that.

> What about tricking mypy? Do you have any idea in mind, and what would be the implications of fooling it just for this cache issue?

"just" this cache is around 5,000 lines of code, so it is kind of a big hit for static type checking ability if we did that, and I'm not even sure I would be able to.

The simpler solution is just to cast to Snowflake yourself when passing stuff to the function imo; we'll do that when we implement frontend methods on the models to interact with the cache anyway.

Up to you if you still want to persue this, though.

**Comment by FasterSpeeding** [comment]
are you going to test each case (type wise) for every method on the cache that takes an ID if this behaviour is added?

**Comment by nekokatt** [comment]
@FasterSpeeding Snowflake inherits the majority of its functionality from int, so it makes no difference.

**Comment by FasterSpeeding** [comment]
Also should this be SlowflakeishOr[Model] or just Snowflakeish

**Comment by nekokatt** [comment]
snowflakeish
i don't want to have to add internal overhead of casting everywhere unnecesarilly when this is an internal component really.

That is why I am still considering closing this.

**Comment by FasterSpeeding** [comment]
> snowflakeish
> i don't want to have to add internal overhead of casting everywhere unnecesarilly when this is an internal component really.
> 
> That is why I am still considering closing this.

would there be much of a difference between the overhead for casting from int to snowflake vs model that supports int to snowflake?

**Comment by FasterSpeeding** [comment]
If you're considering closing this then shouldn't the "priority" tag be removed?

**Comment by nekokatt** [comment]
>>> would there be much of a difference between the overhead for casting from int to snowflake vs model that supports int to snowflake? both should just be as simple as a Snowflake cast right?

int and snowflake are polymorphically compatible, unlike models. 

Closed this anyway.
